agent/consul/rate/handler.go

```diff
@@ -0,0 +1,103 @@
+// package rate implements server-side RPC rate limiting.
+package rate
+
+import (
+	"context"
+	"net"
+	"sync/atomic"
+
+	"github.com/hashicorp/consul/agent/consul/rate/multilimiter"
+)
+
+// Mode determines the action that will be taken when a rate limit has been
+// exhausted (e.g. log and allow, or reject).
+type Mode int
+
+const (
+	// ModePermissive causes the handler to log the rate-limited operation but
+	// still allow it to proceed.
+	ModePermissive Mode = iota
+
+	// ModeEnforce causes the handler to reject the rate-limted operation.
+	ModeEnforce
+)
+
+// OperationType is the type of operation the client is attempting to perform.
+type OperationType int
+
+const (
+	// OperationTypeRead represents a read operation.
+	OperationTypeRead OperationType = iota
+
+	// OperationTypeWrite represents a write operation.
+	OperationTypeWrite
+)
+
+// Operation the client is attempting to perform.
+type Operation struct {
+	// Name of the RPC endpoint (e.g. "Foo.Bar" for net/rpc and "/foo.service/Bar" for gRPC).
+	Name string
+
+	// SourceAddr is the client's (or forwarding server's) IP address.
+	SourceAddr net.Addr
+
+	// Type of operation to be performed (e.g. read or write).
+	Type OperationType
```

Comment:
The RFC mentions keeping a mapping of RPC names to operation type (i.e. read vs write).

I wasn't sure whether we wanted to implement this in the generic handler or in the interceptors themselves, but as the RFC also mentions possibly automating this using protobuf annotations I exposed this on the `Operation` struct incase we handle net/rpc and gRPC methods separately.

We won't be adding any more net/rpc handlers, so I think it's ok to have a static list for these.

Also, with the upcoming controller-based architecture we can implement rate-limiting at the state store which removes this problem entirely (as it's very clear what is a read vs write).